examples/some-issuer/src/bin/discord.rs

```diff
@@ -0,0 +1,397 @@
+    )]
+    issuer_key: PathBuf,
+    #[clap(
+        long = "max-register-energy",
+        help = "The amount of energy to allow for execution of the register credential \
+                transaction. This must be less than max block energy of the chain the service is \
+                connected to.",
+        default_value = "10000",
+        env = "DISCORD_ISSUER_MAX_REGISTER_ENERGY"
+    )]
+    max_register_energy: Energy,
+    #[clap(
+        long = "discord-client-id",
+        help = "Discord client ID for OAuth2.",
+        env = "DISCORD_CLIENT_ID"
+    )]
+    discord_client_id: String,
+    #[clap(
+        long = "discord-client-secret",
+        help = "Discord client secret for OAuth2.",
+        env = "DISCORD_CLIENT_SECRET"
+    )]
+    discord_client_secret: String,
+    #[clap(
+        long = "listen-address",
+        help = "Socket addres for the Discord issuer.",
+        default_value = "0.0.0.0:8081",
+        env = "DISCORD_ISSUER_LISTEN_ADDRESS"
+    )]
+    listen_address: SocketAddr,
+    #[clap(
+        long = "url",
+        help = "URL of the Discord issuer.",
+        default_value = "http://127.0.0.1:8081/",
+        env = "DISCORD_ISSUER_URL"
+    )]
+    url: Url,
+    #[clap(
+        long = "dapp-domain",
+        help = "The domain of the dApp, used for CORS.",
+        default_value = "http://127.0.0.1",
+        env = "DISCORD_ISSUER_DAPP_URL"
+    )]
+    dapp_domain: String,
+}
+
+#[derive(Clone)]
+struct AppState {
+    issuer: IssuerState,
+    discord_client_id: Arc<String>,
+    discord_client_secret: Arc<String>,
+    http_client: reqwest::Client,
+    handlebars: Arc<Handlebars<'static>>,
+    discord_redirect_uri: Arc<Url>,
+    dapp_domain: Arc<String>,
+}
+
+/// Request for issuance of Discord credential.
+#[derive(Debug, Deserialize)]
+struct DiscordIssueRequest {
+    credential: CredentialInfo,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct User {
+    id: String,
+}
+
+#[derive(Serialize)]
+struct AccessTokenRequestData<'a> {
+    client_id: &'a str,
+    client_secret: &'a str,
+    grant_type: &'static str,
+    code: &'a str,
+    redirect_uri: &'a str,
+}
+
+#[derive(Deserialize)]
+#[allow(unused)]
+struct AccessTokenResponse {
+    access_token: String,
+    token_type: String,
+    expires_in: u32,
+    refresh_token: String,
+    scope: String,
+}
+
+#[derive(Serialize)]
+struct OauthTemplateParams<'a> {
+    id: &'a str,
+    dapp_domain: &'a str,
+}
+
+/// Handles OAuth2 redirects and inserts an id in the session.
+async fn handle_oauth_redirect(
+    State(state): State<AppState>,
+    Query(params): Query<HashMap<String, String>>,
+    session: WritableSession,
+) -> Result<Html<String>, StatusCode> {
+    async fn respond(
+        state: AppState,
+        params: HashMap<String, String>,
+        mut session: WritableSession,
+    ) -> anyhow::Result<Html<String>> {
+        let code = params.get("code").context("Missing 'code' query param")?;
```

Comment:
Why is this not handled directly in the type? So instead of
```
Query(HashMap<...>)
```
you'd take a
```
Query(Oauth2RedirectParams)
```
where

```
#[derive(serde::Deserialize)]
struct Oauth2RedirectParams {
   code: String
}
```